### DISCUSSION TITLE
fieldsForDetail is evaluted on Index page by Filters (it should not be)

### DISCUSSION THREAD

**Discussion by cserepesmark** [laravel/nova-issues#5724] [Q&A] [answered]
- Laravel Version: 9.52.10
- Nova Version: 4.26.4
- PHP Version: 8.1.13

### Description:

I have a Product Nova Resource in which I would like to have very separate fields for Index and Details. The reason for this is that I have a very complex custom field which I do not want to show on Index. Nova's Filter function does not respect it.

When you open the Index page of the resource (code) below, you get an Exception on `/nova-api/products/filters`. I think fieldsForDetail should not be run / evaluated on Index page.

### Detailed steps to reproduce the issue on a fresh Nova installation:

```
<?php

namespace App\Nova;

use Laravel\Nova\Fields\Text;
use Laravel\Nova\Http\Requests\NovaRequest;

class Product extends Resource
{
    public static $model = \App\Models\Assets\Product::class;

    public function fields(NovaRequest $request)
    {
        return [];
    }

    public function fieldsForIndex(NovaRequest $request)
    {
        return [
            Text::make('Label'),
        ];
    }

    public function fieldsForDetail(NovaRequest $request)
    {
        throw new \Exception('it was called');
        return [];
    }

    public function fieldsForCreate(NovaRequest $request)
    {
        return $this->fieldsForDetail($request);
    }

    public function fieldsForUpdate(NovaRequest $request)
    {
        return $this->fieldsForDetail($request);
    }
}
```

**Comment by crynobone** [comment, marked as answer]
Unfortunately this is required to support filterable field on BelongsToMany etc where it is specific for Detail .

**Comment by mathieufrh** [comment]
This is an issue for me too. I have a resource with a lot of fields that may take time to compute or retrieve. The resource is read-only. I use fieldsForIndex, fieldsForDetail and fields.
On the index page, I want only a subset of the fields, which are not time-consuming to fetch. But I notice that the index page takes a lot of time to load, most of the time it exceeds the timeout. And it seems to be because the fieldsForDetail and fields methods are evaluated. I try to use `if ($request->isResourceDetailRequest())` to avoid returning all the fields, but I get an error `Call to a member function isShownOnPreview() on null` I don't know how to fix this.

**Reply by cserepesmark** [reply]
Does any of your fields grab data from database? (eg.: select field with options retrived from db)
